### PR TITLE
Add a shutdown reason and rename a lot of things.

### PR DESCRIPTION
**Pull request by de-vri-es**
@dan-jfisher: Can you review?

It's a pretty big API rewrite, so I hope I managed to capture it entirely in the changelog. Have a look at the changelog for an overview of the changes.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,3 +1,19 @@
+# Unreleased:
```
Comment (dan-jfisher) [inline CHANGELOG.md:1]:
```suggestion
# Unreleased:
* Replace `ShutdownAlreadyCompleted` with `ShutdownAlreadyTriggered` which contains both the initial shutdown reason, and the second, ignored, shutdown reason.
```

File: src/wrap_cancel.rs [hunk 2]
```diff
@@ -9,33 +9,40 @@ use crate::shutdown_signal::ShutdownSignal;
...
 
 		// Otherwise check if the shutdown signal has been given.
-		Pin::new(&mut me.shutdown_signal)
-			.poll(context)
-			.map(|()| None)
+		let shutdown = Pin::new(&mut me.shutdown_signal)
+			.poll(context);
+		match shutdown {
+			Poll::Ready(reason) => {
+				me.future = Err(reason.clone());
```
[the remark this answers is not in the document]
Comment (dan-jfisher) [reply src/wrap_cancel.rs:42]: Hmmm, a quick google search shows that `Future` makes no requirements about that behavior, so I guess this is a good practice.

[the remark this answers is not in the document]
Comment (de-vri-es) [reply src/wrap_cancel.rs:42]: well, you're allowed to panic too. We could do that. But I figured that `T` must be clone already, so why not.

File: CHANGELOG.md (hunk 1 above)
Comment (de-vri-es) [reply CHANGELOG.md:1]: But I did forget to list the changes regarding these types.

File: src/wrap_cancel.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (dan-jfisher) [reply src/wrap_cancel.rs:42]: Yeah this behavior is nicer.

**Review by dan-jfisher — APPROVED** [review]
Once the changelog includes the new error type I'm happy

**Pull request merged**